src/main/java/graphql/GraphQL.java

```diff
@@ -289,18 +300,27 @@ public ExecutionResult execute(ExecutionInput executionInput) {
     }
 
     private ExecutionResult parseValidateAndExecute(ExecutionInput executionInput) {
-        ParseResult parseResult = parse(executionInput);
-        if (parseResult.isFailure()) {
-            return toParseFailureExecutionResult(parseResult.getException());
+        PreparsedDocumentEntry preparsedDoc = preparsedDocumentProvider.get(executionInput.getQuery(), query -> {
+            ParseResult parseResult = parse(executionInput);
+            if (parseResult.isFailure()) {
+                return new PreparsedDocumentEntry(toInvalidSyntaxError(parseResult.getException()));
+            } else {
+                final Document document = parseResult.getDocument();
+
+                final List<ValidationError> errors = validate(executionInput, document);
+                if (!errors.isEmpty()) {
+                    return new PreparsedDocumentEntry(errors);
+                }
+
+                return  new PreparsedDocumentEntry(document);
+            }
+        });
+
+        if (preparsedDoc.hasErrors()) {
```

Comment:
It can if the document provider for what ever reason returns null. We could either let it throw a NPE or wrap it into a more descriptive exception.
The lambda will always return a non-null result though.